### PR TITLE
handle retrofit errors similar to feign errors fixes #116

### LINKED ISSUE #116: Retrofit client factory should support propagating exceptions
Currently, any clients constructed through Retrofit will not include any information about the inner exception in the Java RetrofitException that comes back, despite the JSON representation having an extra field that specifically contains such information. This effectively means that certain inner exception information is swallowed.

For Feign clients, it appears we do extra work in FeignClientFactory to support this, likely through our custom SerializableErrorErrorDecoder.java. The request is to do something similar with our Retrofit clients, perhaps by building a Retrofit's custom ErrorHandler.

Examples of exceptions that get swallowed include any errors from Spark, such as exceptions relating to jagged rows in our TextDataFrameReader.

### PR DESCRIPTION
**Pull request by pnepywoda**
@uschi2000

### CODE REVIEW

**Comment by pnepywoda** [comment]
@eshlin

File: retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+
+        InputStream body;
+        try {
+            body = response.getBody().in();
+        } catch (IOException e) {
+            return new RuntimeException("Cannot get input stream from response: " + e.getMessage(), e);
+        }
+        return SerializableErrorErrorDecoder.INSTANCE.decode(contentTypes, response.getStatus(), response.getReason(),
```
Comment (uschi2000) [inline retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java]: move into try block?

**Comment by uschi2000** [comment]
very nice. just to confirm: this doesn't break any compatibility, right?

**Comment by eshlin** [comment]
awesome - looks great. thanks for implementing. +1

File: retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java (hunk 1 above)
Comment (pnepywoda) [inline retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java]: well, the point of the error decoder was to never throw an exception but to always return one....do you still think we should wrap in try/catch though? (the try block above is actually specific to getting the input stream currently....you propose we widen the responsibility of the try/catch?)

**Comment by pnepywoda** [comment]
ping

**Comment by pnepywoda** [comment]
1. made util instead of enum
2. there's no backcompat break

File: retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java (hunk 1 above)
Comment (uschi2000) [inline retrofit-clients/src/main/java/com/palantir/remoting/retrofit/errors/RetrofitSerializableErrorErrorHandler.java]: ok, let's leave it as it is actually.

**Comment by uschi2000** [comment]
build broken, otherwise +1

**Comment by pnepywoda** [comment]
@uschi2000 build isn't broken? can you merge?

**Pull request merged**